api/rest/restapi.go

```diff
@@ -384,8 +385,31 @@ func (api *API) addFileHandler(w http.ResponseWriter, r *http.Request) {
 		sendAcceptedResponse(w, errors.New("unsupported media type"))
 		return
 	}
+	outChan := make(chan *ipld.Node)
+	ctx, cancel := context.WithCancel(api.ctx)
+	go func() {
```

Comment:
Ok, I had a really long comment here until I've figured out that ImportToChannel is probably hanging and that's why we need a a goroutine here to consume the channel.

With the new ImportToChannel signature, I think it should return a channel immediately. It's ImportToChannel implementation's problem to make sure the channel is correctly fed (by launching a goroutine there and managing it with the given context).

The job in this api handler is just to consume that channel (which can then happen in the main thread) and send the response when finished.

Also, I think we do the file streaming using http 1.1. From the go docs:

```
// Depending on the HTTP protocol version and the client, calling
        // Write or WriteHeader may prevent future reads on the
        // Request.Body. For HTTP/1.x requests, handlers should read any
        // needed request body data before writing the response. Once the
        // headers have been flushed (due to either an explicit Flusher.Flush
        // call or writing enough data to trigger a flush), the request body
        // may be unavailable. For HTTP/2 requests, the Go HTTP server permits
        // handlers to continue to read the request body while concurrently
        // writing the response. However, such behavior may not be supported
        // by all HTTP/2 clients. Handlers should read before writing if
        // possible to maximize compatibility.
```
We probably want to stream some sort of information in the response as we are adding (in the future). But it seems we should not be doing that with http1.1 (and use http/2 instead).

Reply:
100% with you.  Im planning to fix this (no goroutine) in PR #318 after this is merged so i can rebase the changes in.  The reason im not doing it now is that I dont want to make changes to /publish the defunct dex repo which id need to do to get the correct ImportTo signature.

👍 on moving to http 2 as well

Reply:
ok!